pkg/guestagent/events/containerd_linux.go

```diff
@@ -0,0 +1,369 @@
+// SPDX-FileCopyrightText: Copyright The Lima Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package events
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/api/events"
+	"github.com/containerd/containerd/errdefs"
+	containerdNamespace "github.com/containerd/containerd/namespaces"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/protobuf/proto"
+
+	"github.com/lima-vm/lima/v2/pkg/guestagent/api"
+)
+
+const (
+	stateKey             = "nerdctl/state-dir"
+	portsKey             = "nerdctl/ports"
+	namespaceKey         = "nerdctl/namespace"
+	defaultSocketTimeout = 5 * time.Second
+)
+
+type ContainerdEventMonitor struct {
+	socketPaths            []string
+	runningContainersMutex sync.Mutex
+	runningContainers      map[string][]*api.IPPort
+}
+
+func NewContainerdEventMonitor(socketPaths []string) *ContainerdEventMonitor {
+	return &ContainerdEventMonitor{
+		socketPaths:       socketPaths,
+		runningContainers: make(map[string][]*api.IPPort),
+	}
+}
+
+func (c *ContainerdEventMonitor) MonitorPorts(ctx context.Context, ch chan *api.Event, errCh chan error) {
+	var wg sync.WaitGroup
+	for _, socket := range c.socketPaths {
+		wg.Add(1)
+		go func(socket string) {
+			defer wg.Done()
+			for {
+				select {
+				case <-ctx.Done():
+					errCh <- ctx.Err()
+					return
+				default:
+				}
+				info, err := os.Stat(socket)
+				if err != nil {
+					if os.IsNotExist(err) {
+						logrus.Warnf("containerd socket %s does not exist", socket)
+						// Wait for 2s before retrying again
+						time.Sleep(2 * time.Second)
```

Comment:
I think we should always sleep for 2 seconds. We should not repeatedly call `os.Stat()` when it returns any error.

This loop also is not cancellable if the error persists. We should use something like this:

```suggestion
						select {
						case <-ctx.Done():	return
						case <-time.After(2 * time.Second):
						}
```